### ISSUE TITLE
Highlighting for string interpolation in Python 3.6 not supported.

### ISSUE THREAD

**Issue by pykong** [ctf0/Seti_UX#22]
**First of all: Thank you for putting up such helpful color scheme, it is a joy to work with!**

Python 3.6 added string literals as an easy way of formatting strings:

    f'He said his name is {name} and he is {age} years old.'

This new feature was well received by the community and therefore will be a much used pattern as more folks migrate to Python 3.6. (You can learn more on string interpolation [here](https://cito.github.io/blog/f-strings/))

Hence it is vital for Seti UX to learn to recognize this new syntax aspect, which it currently does not do accurately or maybe not at all.

Example:

![str_inter](https://user-images.githubusercontent.com/20211140/27879386-b8c9986a-61c1-11e7-9bd3-74e8b48fee6b.png)

Issues in the example:
1) The second and third of the curly brackets are recognized but their content is highlighted in the same color as the bracket themselves / not reflecting the type. (See lines above for correct color. Int - red, Str - blue).
2) Fourth and fifth curly bracket do not appear to be recognized at all. (_Any python expression, even function calls can now be embedded within those curly brackets._)

I assume the bracket highlighting in the current version of Seti UX was made for str.format() and the new string interpolation is not parsed yet at all.

**Comment by ctf0** [comment]
first am happy to find others still enjoy such a scheme :trophy: 
second the high lighting is very much related to the syntax you use, means that if you use a python package to which have much more support than the one that ships with ST , it will be hard for me to keep up with both & thats why i mostly stick to ST default syntax and sadly it doesnt get updated as much and even when it does it add breaking changes.

still as you are more accustomed to python than me, can you make the changes you need and send me a PR ?

**Comment by bfelder** [comment]
Thanks for your fast reply. 
I am not sure whether I will be able to add the changes myself.
I am going to let you know after I had a deeper look into the source of Seti UX on the weekend.

**Comment by bfelder** [comment]
Now that I have checked the regarding xml file out, I think the way forward is to see whether other color schemes support string interpolation and then adopt their approach. I am going to complete that task during next week and will post my results/ pull request.

**Comment by ctf0** [comment]
awesome

**Comment by bfelder** [comment]
Hey ctf0, as I already suspected we would require additional scope definitions to highlight that new syntax. Fortunately other smart guys have already done that work for us. Have a look at the MagicPython package by a fellow sublime lover: [https://github.com/MagicStack/MagicPython/blob/master/grammars/MagicPython.cson](url)
See lines 4059-4104 for the definitions of string interpolations. I do not know how to proceed forward from here, as Seti UX does not contain a similar resource for scope definition. I am sure you have an idea what to do?

**Comment by ctf0** [comment]
sadly it wont work in both default & MagicPython, because of the supported scopes which is found under one and not the other, 
however if u r still into it then get https://packagecontrol.io/packages/ScopeHunter and use it to easily id the scope you want then simply open seti_ux and add that scope with the hl you want

**Comment by bfelder** [comment]
Thanks for the link. ScopeHunter is a nice addition to ColorSchemeEditor, as it allows better display of the scopes. 
Yes, I am still into it. Certainly many of our fellow Sublime Text users would love Seti UX with Python 3.6 support as well. Would you like receive the results of my modification efforts? You can then decide whether to publish those as a python 3.6 compatible fork of the awesome Seti UX scheme.

**Comment by ctf0** [comment]
np at all, send me the PR and will add the new mods it in the release notes 👍

**Comment by bfelder** [comment]
Awesome! It will take me a couple of days though. So be patient. :+1:

**Comment by bfelder** [comment]
I have good news, ctf0: I believe it is possible to create a single thTheme file supporting both Sublime's default python scopes as well as the advanced highlighting features of MagicPython. Also as far as I know there aren't many other color schemes for Sublime supporting the full Python 3.5+ syntax. So, the new Seti UX may become even more popular then ever before!

My problem is that I can not get the color scheme to render: The colors never update. I can not tell whether the problem is with my sublime settings or my theme file. I tried several fixes, including emptying the color_scheme caches of my views. All to no avail.

**Can you have a look into the tmTheme file and check whether you see any obvious mistakes?**
(You are of course much more experienced with Sublime theme editing then me.)
This is not a pull request yet. Colors in that file are totally random as well.

Find the file here: https://github.com/bfelder/Seti_UX/blob/master/Seti.tmTheme
The MagicPython section starts at line: 3149

Tell me if you need any other information.

**Comment by ctf0** [comment]
1- `</string>` should be on the same line as the opening tag
2- if `<key>name</key>` is empty it should be removed
3- updating the scheme, reflects automatically on the opened files so no need for clearing the cache or anything extra

**Comment by bfelder** [comment]
Indeed I totally overlooked the **\n** in the string tag, due to column wrapping in my editor. Now it works thanks. I spent the last couple of hours going through an extensive review and modification of the new scopes' colors, comparing the default with the new version. I have now uploaded an updated version of the tmTheme file. I am going to make a pull request out of it, after I have added  some documentation on the commit.

You can have a peek in case you are impatient. You would require the MagicPython package installed to Sublime. Also MagicPython includes an extensive collection of python test files covering any conceivable syntax scenario. This is great for development. (Find it here: [link](https://github.com/MagicStack/MagicPython))

Note that you should see my current upload as a starting point. There are some colors I am not yet satisfied with. Certainly it takes your design expertise to make the new version of Seti UX both beautiful and still highly legible as we know it.

Shouldn't be too much work still. Just a couple of defined areas requiring our attention. Will come back with more info soon.

**Comment by ctf0** [comment]
https://github.com/ctf0/Seti_UX/pull/23

**Comment by ctf0** [comment]
@bfelder i couldnt stand the fuchsia everywhere, so i changed some colors, plz download the latest commit and check if u can still tell which is which

**Comment by bfelder** [comment]
@ctf0 As I said you are more experienced in choosing colors. :)  Indeed I was not 100% satisfied with the commit I provided. So, I kept improving the scheme. Yet, I was not going to make a PR until further testing. I'd also like to t extend the syntax support to the other two popular python language definitions.
This would make Seti UX stand out as the first color scheme supporting all available Python syntax definitions.

I busy with other things right now. So, do not expect a reply from me sooner than next Sunday.

**Comment by ctf0** [comment]
its okay no rush, take ur time 👍

**Comment by bfelder** [comment]
@ctf0 I am on it. I will open PR once it is ready,

**Comment by bfelder** [comment]
@ctf0 Check by tomorrow. I may have a PR ready by then.

**Comment by bfelder** [comment]
PR is ready. Check if you like it. Fuchsia should be much less prevalent by this iteration.
Still, I am very unsatisfied how regexp is displayed in MagicPython. For some reason, any changes I made to the tmTheme file did not punch through to the actual display. As the colors have been randomly generated for my first PR, the result is very unreadable.

Also, note that I am going to include definitions for [Python 3](https://packagecontrol.io/packages/Python%203) and [Python Improved](https://packagecontrol.io/packages/Python%20Improved). I did not include those in the current PR, as it is best to first establish a good color language and then extending it to other syntaxes.
